datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -500,302 +387,386 @@ fn optimize_join(
+                let mut push_predicates = vec![];
+                for expr in predicates {
+                    let columns = expr.to_columns()?;
+                    if columns.is_empty()
+                        || !columns
+                            .intersection(&used_columns)
+                            .collect::<HashSet<_>>()
+                            .is_empty()
+                    {
+                        keep_predicates.push(expr);
+                    } else {
+                        push_predicates.push(expr);
                     }
-                    new_filters.push(filter_expr.clone());
+                }
+
+                let child = match conjunction(push_predicates) {
+                    Some(predicate) => LogicalPlan::Filter(Filter::try_new(
+                        predicate,
+                        Arc::new((*window.input).clone()),
+                    )?),
+                    None => (*window.input).clone(),
+                };
+                let new_agg = from_plan(
+                    filter.input(),
+                    &filter.input().expressions(),
+                    &vec![child],
+                )?;
+                match conjunction(keep_predicates) {
+                    Some(predicate) => LogicalPlan::Filter(Filter::try_new(
+                        predicate,
+                        Arc::new(new_agg),
+                    )?),
+                    None => new_agg,
+                }
+            }
+            LogicalPlan::Join(join) => {
+                match push_down_join(filter.input(), join, Some(filter.predicate()))? {
+                    Some(optimized_plan) => optimized_plan,
+                    None => plan.clone(),
                 }
             }
+            LogicalPlan::CrossJoin(CrossJoin { left, right, .. }) => {
+                let predicates = utils::split_conjunction_owned(utils::cnf_rewrite(
+                    filter.predicate().clone(),
+                ));
 
-            issue_filters(
-                state,
-                used_columns,
-                &LogicalPlan::TableScan(TableScan {
-                    source: source.clone(),
-                    projection: projection.clone(),
-                    projected_schema: projected_schema.clone(),
-                    table_name: table_name.clone(),
-                    filters: new_filters,
-                    fetch: *fetch,
-                }),
-            )
-        }
-        _ => {
-            // all other plans are _not_ filter-commutable
-            let used_columns = plan
-                .schema()
-                .fields()
-                .iter()
-                .map(|f| f.qualified_column())
-                .collect::<HashSet<_>>();
-            issue_filters(state, used_columns, plan)
-        }
-    }
-}
+                push_down_all_join(predicates, filter.input(), left, right, vec![])?
+            }
+            LogicalPlan::TableScan(scan) => {
```

Comment:
I have a feeling that the data source/table scan filter push down should be handled in a physical planing phases, either by a physical rule or during converting LogicalPlan::TableScan to a physical Exec(ParquetExec).

Reply:
Is the suggestion here that the `LogicalPlan::TableScan` will no longer contain filters at the logical planning phase. In that case would this rule just generate filters that can be pushed down to be used by some followup rule? 
(Apologies if the question doesn't make too much sense, I'm still not completely familiar with datafusion internals especially around physical plans).

Reply:
Yes, that was my understanding of the proposal -- however I think the merits of that change should be discussed in a different Issue / PR -- this one has gotten crazy already!